Upgrade to 15.0.0

mamtashukla opened the pull request:
This series of patch allows to build compiler-rt for 15.0.0 version:
1d75a0a patches: remove patch for __is_convertible
f172b04 patches: add patch to remove abi-tag
a5178e9 Makefile.uk: Use -fpermissive flag
8ae7efe Makefile.uk: Upgrade to 15.0.0

Fixes: https://github.com/unikraft/lib-compiler-rt/issues/7

mamtashukla committed:
Makefile.uk: Upgrade to 15.0.0

Signed-off-by: Mamta Shukla <mamtashukla555@gmail.com>

mamtashukla committed:
patches: Add patch to remove abi-tag

ABI_TAG was introduced in upstream:
https://github.com/llvm/llvm-project/commit/67b0b02ec9f2bbc57bf8f0550828d97f460ac11f

These ABI_TAG leads to error:

build/libcxx/origin/libcxx-15.0.0.src/include/__support/musl/xlocale.h:38:74:
error: ‘abi_tag’ attribute applied to extern "C" declaration ‘long long
int wcstoll_’
   38 | wcstoll_l(const wchar_t *__nptr, wchar_t **__endptr, int __base,
locale_t) {

Signed-off-by: Mamta Shukla <mamtashukla555@gmail.com>

mamtashukla committed:
patches: Remove patch for __is_convertible

__is_convertible patch does not applies to 15.0.0 version.

It is possible to compile libcxx without this patch. Tested for
gcc-11.

This reverts the commit:
4dd2450604: patches: Backport upstream fix for __is_convertible

Signed-off-by: Mamta Shukla <mamtashukla555@gmail.com>

mariasfiraiala reviewed (changes requested):
Some observations @mamtashukla:

1. The changes compile with `gcc-11` **only** if the `CXX_THREADS` option is enabled. The compile errors look something like this:

 ```Makefile
In file included from /home/maria/arm/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-15.0.0.src/src/cxa_guard.cpp:15:
/home/maria/arm/apps/app-helloworld-cpp/build/libcxxabi/origin/libcxxabi-15.0.0.src/src/cxa_guard_impl.h:288:8: error: ‘__libcpp_mutex_t’ in namespace ‘std’ does not name a type
288 |   std::__libcpp_mutex_t mutex = _LIBCPP_MUTEX_INITIALIZER;
  |        ^~~~~~~~~~~~~~~~
```

There are a multitude of errors regarding mutexes, conditional variables etc. My guess is that this is due to the placement of the libs in the Makefile: 

   ```Makefile
   LIBS := $(UK_LIBS)/lib-libcxxabi:$(UK_LIBS)/lib-libcxx:$(UK_LIBS)/lib-compiler-rt:$(UK_LIBS)/lib-libunwind:$(UK_LIBS)/lib-musl
   ```

Basically, AFAIK, C++ mutexes and varconds need pthread support, and musl being compiled the last doesn't help with that. I don't know how to fix this, as moving musl in front of libcxx brings other compiler errors :smiling_face_with_tear: 

2.  For the patch commits please use capital letter after the selector, eg: "patches: **A**dd patch to remove abi-tag"

3. Rename your newly added patch with 0001-*, since you are deleting the old one with this tag.

**LE: The first point is now redundant, as the error is part of `libcxxabi`, and I've already suggested some changes there to fix it.** @mamtashukla you still need to address the last 2 points :smile:

StefanJum reviewed (changes requested):
Thank you @mamtashukla.
Please rename the patches (so we have `0001` first, even if you remove the current one).
Also please use git `format-patch` to create the patches. You should clone the origin `libcxx` repository, do the changes from the patch, make a commit and then run `format-patch`, you would end up with something similar to the patches [here](https://github.com/unikraft/lib-musl/blob/staging/patches/0001-Remove-prototype-for-syscall-from-unistd.h.patch)

mamtashukla commented:
v2:
-Use capital letter for commit after selector
08e75d2  patches: Remove patch for __is_convertible
ccd8b56 patches: Add patch to remove abi-tag

-Fix patch format

mariasfiraiala reviewed (approved):
All good on my side, now, thanks for the prompt update @mamtashukla!

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

StefanJum reviewed (changes requested):
There are some conflicts @mamtashukla, can you please rebase this on top of `staging`?

mamtashukla committed:
Makefile.uk: Use -fpermissive flag

Use -fpermissive flag to allow compiler-rt build

note: (if you use ‘-fpermissive’, G++ will accept your code, but
allowing the use of an undeclared name is deprecated)

build/libcxx/origin/libcxx-15.0.0.src/include/__type_traits/is_same.h:22:72:
error: template argument 1 is invalid
   22 | struct _LIBCPP_TEMPLATE_VIS is_same :
_BoolConstant<__is_same(_Tp, _Up)> { };

Signed-off-by: Mamta Shukla <mamtashukla555@gmail.com>

mamtashukla commented:
> There are some conflicts @mamtashukla, can you please rebase this on top of `staging`?

Rebased on top of staging and tested with a rebuild.
